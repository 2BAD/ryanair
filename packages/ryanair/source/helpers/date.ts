import { addDays, addMonths, eachMonthOfInterval, format, isAfter } from 'date-fns'
import type { StrDate } from '~/date.types.ts'

/**
 * Calculates the date for the next day (formatted as 'YYYY-MM-DD')
 */
export const tomorrow = (): string => {
  const currentDate = new Date()
  const tomorrowDate = addDays(currentDate, 1)

  // Format the date into 'YYYY-MM-DD' format
  return format(tomorrowDate, 'yyyy-MM-dd')
}

/**
 * Calculates the date for the next month (formatted as 'YYYY-MM-DD')
 */
export const nextMonth = (): string => {
  const currentDate = new Date()
  const nextMonthDate = addMonths(currentDate, 1)

  // Format the date into 'YYYY-MM-DD' format
  return format(nextMonthDate, 'yyyy-MM-dd')
}

/**
 * Returns true if the first ISO-formatted date is after the second ISO-formatted date.
 *
 * @param firstDate - The first date (formatted as 'YYYY-MM-DD')
 * @param secondDate - The second date (formatted as 'YYYY-MM-DD')
 */
export const isAfterIso = (firstDate: string, secondDate: string): boolean => {
  const first = new Date(firstDate)
  const second = new Date(secondDate)
  return isAfter(first, second)
}

/**
 * Returns an array of the first day of each month in a given date range (formatted as 'YYYY-MM-01')
 *
 * @param startDate - The start date of the range (formatted as 'YYYY-MM-DD')
 * @param endDate - The end date of the range (formatted as 'YYYY-MM-DD')
 * @throws {RangeError} - Start date must be before the end date.
 */
export const getFirstDayOfEachMonthInRange = (startDate: StrDate, endDate: StrDate): StrDate[] => {
  const start = new Date(startDate)
  const end = new Date(endDate)

  if (start > end) {
    throw new RangeError('Invalid date range. The start date must be before the end date.')
  }

  return eachMonthOfInterval({ start, end }).map((d) => format(d, 'yyyy-MM-01'))
}
