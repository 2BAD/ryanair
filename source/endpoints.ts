export const BASE_API = 'https://www.ryanair.com/api'
export const BOOKING_API = `${BASE_API}/booking/v4/en-gb`
export const FARE_FINDER_API = `${BASE_API}/farfnd/v4`
export const GEOLOCATION_API = `${BASE_API}/geoloc/v5`
export const TIMETABLE_API = `${BASE_API}/timtbl/3`
export const VIEWS_API = `${BASE_API}/views/locate`
