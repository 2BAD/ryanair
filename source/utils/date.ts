import { addDays, addMonths, eachMonthOfInterval, format } from 'date-fns'
import { type StrDate } from '~/date.types'

/**
 * Calculates the date for the next day (formatted as 'YYYY-MM-DD')
 */

export const tomorrow = (): string => {
  const currentDate = new Date()
  const tomorrowDate = addDays(currentDate, 1)

  // Format the date into 'YYYY-MM-DD' format
  return format(tomorrowDate, 'yyyy-MM-dd')
}

/**
 * Calculates the date for the next month (formatted as 'YYYY-MM-DD')
 */

export const nextMonth = (): string => {
  const currentDate = new Date()
  const nextMonthDate = addMonths(currentDate, 1)

  // Format the date into 'YYYY-MM-DD' format
  return format(nextMonthDate, 'yyyy-MM-dd')
}

/**
 * Returns an array of the first day of each month in a given date range (formatted as 'YYYY-MM-01')
 *
 * @param startDate - The start date of the range (formatted as 'YYYY-MM-DD')
 * @param endDate - The end date of the range (formatted as 'YYYY-MM-DD')
 *
 */
export const getFirstDayOfEachMonthInRange = (startDate: StrDate, endDate: StrDate): StrDate[] => {
  const dates = eachMonthOfInterval({ start: new Date(startDate), end: new Date(endDate) })
  return dates.map((d) => format(d, 'yyyy-MM-01'))
}
