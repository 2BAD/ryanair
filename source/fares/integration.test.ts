import * as client from '~/client/index.ts'
import { FARE_FINDER_API } from '~/endpoints.ts'
import { findCheapestRoundTrip, findDailyFaresInRange, getCheapestPerDay } from '~/fares/index.ts'
import { isAfterISO, nextMonth, tomorrow } from '~/helpers/date.ts'

describe('fares', () => {
  describe('getCheapestPerDay', () => {
    it('when provided with all parameters \n\t Then should call the correct API URL', async () => {
      expect.assertions(1)
      const getSpy = vi.spyOn(client, 'get')
      const from = 'BER' // Berlin airport
      const to = 'KRK' // Krakow airport
      const startDate = tomorrow()
      const currency = 'EUR'
      await getCheapestPerDay(from, to, startDate, currency)

      expect(getSpy).toHaveBeenCalledWith(
        `${FARE_FINDER_API}/oneWayFares/${from}/${to}/cheapestPerDay?outboundMonthOfDate=${startDate}&currency=${currency}`
      )
    })

    it('when provided with all parameters \n\t Then should be able to retrieve data and parse it', async () => {
      expect.assertions(1)
      const from = 'BER' // Berlin airport
      const to = 'KRK' // Krakow airport
      const startDate = tomorrow()
      const currency = 'EUR'

      const data = await getCheapestPerDay(from, to, startDate, currency)
      expect(data.outbound.fares.length).toBeGreaterThan(0)
    })

    it('when asked for info on non existing airport \n\t Then should throw HTTP error', async () => {
      const from = 'WRONG_IATA_CODE'
      const to = 'KRK' // Krakow airport
      const startDate = tomorrow()
      const currency = 'EUR'

      await expect(getCheapestPerDay(from, to, startDate, currency)).rejects.toThrow('Response code 400 (Bad Request)')
    })
  })

  describe('findDailyFaresInRange', () => {
    it('when provided with all parameters \n\t Then should call the correct API URL', async () => {
      expect.assertions(1)
      const getSpy = vi.spyOn(client, 'get')
      const from = 'BER' // Berlin airport
      const to = 'KRK' // Krakow airport
      const startDate = tomorrow()
      const endDate = tomorrow()
      const currency = 'EUR'

      const [year, month] = startDate.split('-')
      const firstDayOfMonth = `${year}-${month}-01`

      await findDailyFaresInRange(from, to, startDate, endDate, currency)

      expect(getSpy).toHaveBeenCalledWith(
        `${FARE_FINDER_API}/oneWayFares/${from}/${to}/cheapestPerDay?outboundMonthOfDate=${firstDayOfMonth}&currency=${currency}`
      )
    })

    it('when provided with all parameters \n\t Then should be able to retrieve data and parse it', async () => {
      expect.assertions(1)
      const from = 'BER' // Berlin airport
      const to = 'KRK' // Krakow airport
      const startDate = tomorrow()
      const endDate = nextMonth()
      const currency = 'EUR'

      const data = await findDailyFaresInRange(from, to, startDate, endDate, currency)
      expect(data.length).toBeGreaterThan(0)
    })

    it('when provided with all parameters \n\t Then should not have any nullish prices', async () => {
      expect.assertions(1)
      const from = 'BER' // Berlin airport
      const to = 'KRK' // Krakow airport
      const startDate = tomorrow()
      const endDate = nextMonth()
      const currency = 'EUR'

      const data = await findDailyFaresInRange(from, to, startDate, endDate, currency)
      expect(data.every((fare) => fare.price !== null)).toBeTruthy()
    })
  })

  describe('findCheapestRoundTrip', () => {
    it('when provided with all parameters \n\t Then should call the correct API URL', async () => {
      expect.assertions(1)
      const getSpy = vi.spyOn(client, 'get')
      const from = 'BER' // Berlin airport
      const to = 'KRK' // Krakow airport
      const startDate = tomorrow()
      const endDate = tomorrow()
      const currency = 'EUR'

      const [year, month] = startDate.split('-')
      const firstDayOfMonth = `${year}-${month}-01`

      await findCheapestRoundTrip(from, to, startDate, endDate, currency)

      expect(getSpy).toHaveBeenCalledWith(
        `${FARE_FINDER_API}/oneWayFares/${from}/${to}/cheapestPerDay?outboundMonthOfDate=${firstDayOfMonth}&currency=${currency}`
      )
    })

    it('when provided with all parameters \n\t Then should be able to retrieve data and parse it', async () => {
      expect.assertions(1)
      const from = 'BER' // Berlin airport
      const to = 'KRK' // Krakow airport
      const startDate = tomorrow()
      const endDate = nextMonth()
      const currency = 'EUR'

      const data = await findCheapestRoundTrip(from, to, startDate, endDate, currency)
      expect(data.length).toBeGreaterThan(0)
    })

    it('when provided with all parameters \n\t Then should not have any nullish prices', async () => {
      expect.assertions(2)
      const from = 'BER' // Berlin airport
      const to = 'KRK' // Krakow airport
      const startDate = tomorrow()
      const endDate = nextMonth()
      const currency = 'EUR'

      const data = await findCheapestRoundTrip(from, to, startDate, endDate, currency)
      expect(data.every((trip) => trip.departure.price !== null)).toBeTruthy()
      expect(data.every((trip) => trip.return.price !== null)).toBeTruthy()
    })

    it('when provided with all parameters \n\t Then should have return trips after departure', async () => {
      expect.assertions(1)
      const from = 'BER' // Berlin airport
      const to = 'KRK' // Krakow airport
      const startDate = tomorrow()
      const endDate = nextMonth()
      const currency = 'EUR'

      const data = await findCheapestRoundTrip(from, to, startDate, endDate, currency)
      expect(data.every((trip) => isAfterISO(trip.return.day, trip.departure.day))).toBeTruthy()
    })
  })
})
