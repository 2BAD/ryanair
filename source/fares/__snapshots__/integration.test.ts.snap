// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Fares > cheapestPerDay > When provided with all parameters 
	 Then should be able to retrieve data and parse it 1`] = `
{
  "outbound": {
    "fares": [
      {
        "arrivalDate": "2024-01-01T23:55:00",
        "day": "2024-01-01",
        "departureDate": "2024-01-01T22:45:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 57.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "57",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-02T23:30:00",
        "day": "2024-01-02",
        "departureDate": "2024-01-02T22:20:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.19,
          "valueFractionalUnit": "19",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-03T12:20:00",
        "day": "2024-01-03",
        "departureDate": "2024-01-03T11:10:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 54.94,
          "valueFractionalUnit": "94",
          "valueMainUnit": "54",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-04T23:25:00",
        "day": "2024-01-04",
        "departureDate": "2024-01-04T22:15:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 41.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "41",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-05T08:50:00",
        "day": "2024-01-05",
        "departureDate": "2024-01-05T07:40:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 41.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "41",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-06T22:05:00",
        "day": "2024-01-06",
        "departureDate": "2024-01-06T20:55:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-07T08:20:00",
        "day": "2024-01-07",
        "departureDate": "2024-01-07T07:10:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-08T23:55:00",
        "day": "2024-01-08",
        "departureDate": "2024-01-08T22:45:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-09T23:30:00",
        "day": "2024-01-09",
        "departureDate": "2024-01-09T22:20:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 25.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "25",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-10T12:20:00",
        "day": "2024-01-10",
        "departureDate": "2024-01-10T11:10:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-11T08:30:00",
        "day": "2024-01-11",
        "departureDate": "2024-01-11T07:20:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-12T23:15:00",
        "day": "2024-01-12",
        "departureDate": "2024-01-12T22:05:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-13T23:40:00",
        "day": "2024-01-13",
        "departureDate": "2024-01-13T22:30:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-14T08:20:00",
        "day": "2024-01-14",
        "departureDate": "2024-01-14T07:10:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-15T23:55:00",
        "day": "2024-01-15",
        "departureDate": "2024-01-15T22:45:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-16T23:30:00",
        "day": "2024-01-16",
        "departureDate": "2024-01-16T22:20:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 25.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "25",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-17T12:20:00",
        "day": "2024-01-17",
        "departureDate": "2024-01-17T11:10:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-18T08:30:00",
        "day": "2024-01-18",
        "departureDate": "2024-01-18T07:20:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-19T23:15:00",
        "day": "2024-01-19",
        "departureDate": "2024-01-19T22:05:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-20T22:05:00",
        "day": "2024-01-20",
        "departureDate": "2024-01-20T20:55:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-21T08:20:00",
        "day": "2024-01-21",
        "departureDate": "2024-01-21T07:10:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-22T23:55:00",
        "day": "2024-01-22",
        "departureDate": "2024-01-22T22:45:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 41.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "41",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-23T23:30:00",
        "day": "2024-01-23",
        "departureDate": "2024-01-23T22:20:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 25.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "25",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-24T12:20:00",
        "day": "2024-01-24",
        "departureDate": "2024-01-24T11:10:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-25T08:30:00",
        "day": "2024-01-25",
        "departureDate": "2024-01-25T07:20:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-26T12:00:00",
        "day": "2024-01-26",
        "departureDate": "2024-01-26T10:55:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 41.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "41",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-27T23:40:00",
        "day": "2024-01-27",
        "departureDate": "2024-01-27T22:30:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 41.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "41",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-28T23:50:00",
        "day": "2024-01-28",
        "departureDate": "2024-01-28T22:40:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 41.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "41",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-29T23:55:00",
        "day": "2024-01-29",
        "departureDate": "2024-01-29T22:45:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 41.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "41",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-30T23:30:00",
        "day": "2024-01-30",
        "departureDate": "2024-01-30T22:20:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 25.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "25",
        },
        "soldOut": false,
        "unavailable": false,
      },
      {
        "arrivalDate": "2024-01-31T12:20:00",
        "day": "2024-01-31",
        "departureDate": "2024-01-31T11:10:00",
        "price": {
          "currencyCode": "EUR",
          "currencySymbol": "€",
          "value": 30.99,
          "valueFractionalUnit": "99",
          "valueMainUnit": "30",
        },
        "soldOut": false,
        "unavailable": false,
      },
    ],
    "maxFare": {
      "arrivalDate": "2024-01-01T23:55:00",
      "day": "2024-01-01",
      "departureDate": "2024-01-01T22:45:00",
      "price": {
        "currencyCode": "EUR",
        "currencySymbol": "€",
        "value": 57.99,
        "valueFractionalUnit": "99",
        "valueMainUnit": "57",
      },
      "soldOut": false,
      "unavailable": false,
    },
    "minFare": {
      "arrivalDate": "2024-01-09T23:30:00",
      "day": "2024-01-09",
      "departureDate": "2024-01-09T22:20:00",
      "price": {
        "currencyCode": "EUR",
        "currencySymbol": "€",
        "value": 25.99,
        "valueFractionalUnit": "99",
        "valueMainUnit": "25",
      },
      "soldOut": false,
      "unavailable": false,
    },
  },
}
`;
